artemis-commons/src/main/java/org/apache/activemq/artemis/utils/critical/CriticalMeasure.java

```diff
@@ -17,28 +17,39 @@
 
 package org.apache.activemq.artemis.utils.critical;
 
-import org.jboss.logging.Logger;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLongFieldUpdater;
 
-public class CriticalMeasure {
+final class CriticalMeasure {
 
-   private static final Logger logger = Logger.getLogger(CriticalMeasure.class);
+   //uses updaters to avoid creates many AtomicLong instances
+   private static final AtomicLongFieldUpdater<CriticalMeasure> TIME_ENTER_UPDATER = AtomicLongFieldUpdater.newUpdater(CriticalMeasure.class, "timeEnter");
+   private static final AtomicLongFieldUpdater<CriticalMeasure> TIME_LEFT_UPDATER = AtomicLongFieldUpdater.newUpdater(CriticalMeasure.class, "timeLeft");
 
-   private volatile long timeEnter;
-   private volatile long timeLeft;
+   //System::nanoTime can't reach this value so it's the best candidate to have a NULL semantic
+   private static final long NIL = Long.MAX_VALUE;
```

Comment:
just set it to zero like we had before...


and all you need is if (timeEnter > timeLeft)

Reply:
Can't be done sadly.
It is mostly OS dependent and on Linux nanoTime can start with Long.MIN too (or near to it). So having it with a safe "null" as I've done is the only way to let it work across OSes.

Reply:
I can do an approximation of the original logic by initializing both the variables to Long.MIN in order to maintain this logic simple as the original one and to cover the case where only timeLeft is not initialized with nanoTime calls negative.